src/main/java/com/microsoft/aad/adal4j/SSLProxyTunnelSocketFactory.java

```diff
@@ -0,0 +1,192 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.adal4j;
+
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+/**
+ * Supports proxy tunneling with authentication.
+ */
+public class SSLProxyTunnelSocketFactory extends SSLSocketFactory {
+    private SSLSocketFactory defaultFactory;
+
+    private String tunnelHost;
+
+    private int tunnelPort;
+
+    private ChallengeHandler challengeHandler;
+
+    /**
+     * Creates an SSLProxyTunnelSocketFactory for proxy tunneling.
+     *
+     * @param proxyHost the host of the proxy
+     * @param proxyPort the port of the proxy
+     * @param challengeHandler the handler to handle authentication challenges
+     */
+    public SSLProxyTunnelSocketFactory(String proxyHost, int proxyPort, ChallengeHandler challengeHandler) {
+        tunnelHost = proxyHost;
+        tunnelPort = proxyPort;
+        defaultFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
+
+        this.challengeHandler = challengeHandler;
+    }
+
+    public Socket createSocket(String host, int port) throws IOException {
+        return createSocket(null, host, port, true);
+    }
+
+    public Socket createSocket(String host, int port, InetAddress clientHost,
+                               int clientPort) throws IOException {
+        return createSocket(null, host, port, true);
+    }
+
+    public Socket createSocket(InetAddress host, int port) throws IOException {
+        return createSocket(null, host.getHostName(), port, true);
+    }
+
+    public Socket createSocket(InetAddress address, int port,
+                               InetAddress clientAddress, int clientPort) throws IOException {
+        return createSocket(null, address.getHostName(), port, true);
+    }
+
+    public Socket createSocket(Socket s, String host, int port,
```

Comment:
Implementation is not aligned with exposed API (where you choose proxy Authentication type explicitly).
If you choose Basic Authentication type, and corresponding handler will be used - what is the purpose of  issuing "CONNECT" request to get challenge ? since challenge is required to digest only That call will be redundant.

If you want to simplify API and remove type of proxy authentication from public setProxy api , you will have to 
issue "CONNECT" request to figure out type of Authentication used by proxy.

Reply:
Thanks for the insights.
I could not find information on whether challenge based systems can also use Basic auth thus implemented this way. 
I did not used `java.net.Authenticator` for the reason of avoiding JVM wide configuration. This implementation opens up for a big variety of authentication mechanisms as long as a challenge can be satisfied. 

My next steps:
1) Address the CR feedback above (thank you!)
2) Support rfc7616
3) Add unit tests for digest algorithm
4) Modify the exposed methods to allow accessing credentials from `java.net.Authenticator` by default
5) In Basic auth scenario, I will send the header on the first CONNECT to proxy server. I'll have to change my challenge interface to allow this so hopefully it won't look too ugly.

Please advise.